sycl/doc/SpecializationConstants.md

```diff
@@ -0,0 +1,305 @@
+the `__sycl_getSpecConstantValue` calls with constants - default values of
+the spec constant's type. No maps are generated, and SYCL program can't change
+the value of a spec constant.
+
+#### LLVMIR-SPIR-V translator
+
+Given the `__spirv_SpecConstant` intrinsic calls produced by the
+`SpecConstants` pass:
+```cpp
+; Function Attrs: alwaysinline
+define dso_local spir_func i32 @get() local_unnamed_addr #0 {
+  ; args are "ID" and "default value":
+  %1 = tail call spir_func i32 @_Z20__spirv_SpecConstantii(i32 42, i32 0)
+  ret i32 %1
+}
+```
+
+the translator will generate `OpSpecConstant` SPIR-V instructions with proper
+`SpecId` decorations:
+
+```cpp
+              OpDecorate %i32 SpecId 42 ; ID
+       %i32 = OpSpecConstant %int 0     ; Default value
+         %1 = OpTypeFunction %int
+
+       %get = OpFunction %int None %1
+         %2 = OpLabel
+              OpReturnValue %i32
+              OpFunctionEnd
+```
+
+#### SYCL runtime
+
+For each device binary compiler generates a map \<Symbolic ID\> =\> \<numeric ID\>
+("ID map"). The SYCL runtime imports that map when loading device binaries.
+It also maintains another map \<Spec const symbolic ID\> =\> \<its value\> 
+("value map") per `sycl::program` object. The value map is updated upon
+`program::set_spec_constant<IDType>(val)` calls from the app.
+
+**_NOTE_**  `IDType` gets translated to the symbolic ID using the integration
+header mechanism, similarly to kernel ID type. The reason why
+`__builtin_unique_stable_name` is not used here is because this code is
+compiled by the host compiler, which can be any C++ 14-compatible compiler
+unaware of the clang-specific built-ins.
+
+Before JIT-ing a program, the runtime "flushes" the spec constants: it iterates
+through the value map and invokes the
+
+```cpp
+pi_result piextProgramSetSpecializationConstant(pi_program prog,
+                                                pi_uint32 spec_id,
+                                                size_t spec_size,
+                                                const void *spec_value);
+```
+
+Plugin Interface function for each entry, taking the `spec_id` from the ID map.
+
+## "Plain Old Data" (POD) types support design
+
+#### Source representation
+
+Say, the POD type is
+
+```cpp
+struct A {
+  int x;
+  float y;
+};
+
+struct POD {
+  A a[2];
+  int b;
+};
+```
+
+and the user says
+
+```cpp
+  POD gold{
+    {
+      { goldi, goldf },
+      { goldi + 1, goldf + 1 },
+    },
+    goldi
+  };
+
+  cl::sycl::ONEAPI::experimental::spec_constant<POD, MyConst> sc =  program4.set_spec_constant<MyConst>(gold);
+```
+
+#### Compiler
+
+##### The SpecConstant pass changes
+
+ - The SpecConstants pass in the post-link will have the following IR as input (`sret` conversion is omitted for clarity):
+
+```cpp
+  %spec_const = call %struct.POD __sycl_getCompositeSpecConstantValue<mangling for POD type template specialization ("MyConst_mangled")
```

Comment:
```suggestion
  %spec_const = call %struct.POD __sycl_getCompositeSpecConstantValue<mangling for POD type template specialization> ("MyConst_mangled")
```

Reply:
this  just a pseudo-code to illustrate the idea